e2e/sink/internal/handler/handler.go

```diff
@@ -0,0 +1,135 @@
+package handler
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/gorilla/mux"
+	"go.uber.org/zap"
+
+	cev2event "github.com/cloudevents/sdk-go/v2/event"
+	cev2http "github.com/cloudevents/sdk-go/v2/protocol/http"
+)
+
+const (
+	SampledHeader     = "X-B3-Sampled"
+	TraceparentHeader = "Traceparent"
+)
+
+// Handler interface for the SinkHandler.
+type Handler interface {
+	Start() error
+}
+
+type SinkHandler struct {
+	logger *zap.Logger
+	events map[string]*sinkEvent
+}
+
+func NewSinkHandler(logger *zap.Logger) *SinkHandler {
+	return &SinkHandler{
+		logger: logger,
+		events: make(map[string]*sinkEvent),
+	}
+}
+
+func (h *SinkHandler) Start(port string) error {
+	router := mux.NewRouter()
+	router.HandleFunc("/events", h.StoreEvent).Methods(http.MethodPost)
+	router.HandleFunc("/events/{eventID}", h.GetEvent).Methods(http.MethodGet)
+
+	return http.ListenAndServe(":"+port, router)
+}
+
+type sinkEvent struct {
+	// Header stores the non CE events, e.g. X-B3-Sampled and Traceparent
+	http.Header
+	cev2event.Event
+}
+
+func (h *SinkHandler) StoreEvent(w http.ResponseWriter, r *http.Request) {
+	event, err := extractCloudEventFromRequest(r)
+	if err != nil {
+		h.namedLogger().With().Error("failed to extract CloudEvent from request", zap.Error(err))
+		e := writeResponse(w, http.StatusBadRequest, []byte(err.Error()))
+		if e != nil {
+			h.namedLogger().Error("failed to write response", zap.Error(e))
+		}
+		return
+	}
+
+	// store the event in memory
+	h.events[event.ID()] = &sinkEvent{
+		Header: http.Header{
+			SampledHeader:     r.Header[SampledHeader],
+			TraceparentHeader: r.Header[TraceparentHeader],
```

Comment:
Question: Will it fail here or return empty string if these headers are missing in the request?

Reply:
no, it just doesn't set it. I've tested it.